src/common/main.ts

```diff
@@ -107,3 +107,35 @@ export function noop(): void {
 export function isFloat(n: any): boolean {
     return Number(n) === n && n % 1 !== 0;
 }
+
+export function isObject(item: any): boolean {
+    return (item && typeof item === 'object' && !Array.isArray(item));
+}
+
+// Deep merge https://stackoverflow.com/a/34749873
+export function mergeDeep(target: any, ...sources: any[]): any {
```

Comment:
we also do this [here](https://github.com/HadoukenIO/core/blob/develop/src/browser/convert_options.js#L151). If we are doing the same thing in 2 places lets use the same functions to do it.

Reply:
The function you mentioned (`validate`) does these things differently:
1. **Important difference**: it only sets overwrites base values if both types of base & new values match. If they don't, it'll keep the base value.
2. **Minor difference**: It also only checks if the value is an object, but it could be an array. I know we don't have arrays in windows options, but...

Reply:
For 1, isn't that what we want? do we have cases where the thing you set it to is a different type than the default?
For 2, I'd argue that merging an array is not a thing we ever want to do

Reply:
For 1, I'm afraid switching that method to use newly added `deepMerge`, because deepMerge will simply overwrite provided values. In this function, we only do that if their types match. It sounds risky to me because they slightly do different things.

Reply:
why not use that method in place of deep merge?

Reply:
I'm not sure about all possible scenarios of updating window options. For example, we used to have preloadScripts (and plugins, I think) with different value types, and it wouldn't work in those cases where there are different value types. So if we re-use current method, we would have to let users know that the options are not being "extended, only for matching types.